MainWindow.cpp

```diff
@@ -105,34 +108,23 @@ void MainWindow::openSubWindow(buffers::TreeNode *item) {
                                 {TypeCase::kObject, EditorFactory<ObjectEditor>},
                                 {TypeCase::kRoom, EditorFactory<RoomEditor>}});
 
-  const Descriptor *desc = item->GetDescriptor();
-  const Reflection *refl = item->GetReflection();
-
-  const OneofDescriptor *typeOneof = desc->FindOneofByName("type");
-  const FieldDescriptor *typeField = refl->GetOneofFieldDescriptor(*item, typeOneof);
-  // might not be set or it's not a message
-  if (!typeField || typeField->cpp_type() != FieldDescriptor::CppType::CPPTYPE_MESSAGE) return;
-  Message *typeMessage = refl->MutableMessage(item, typeField);
-
   auto subWindow = subWindows[item];
   if (!subWindows.contains(item) || subWindow == nullptr) {
     auto factoryFunction = factoryMap.find(item->type_case());
     if (factoryFunction == factoryMap.end()) return;  // no registered editor
 
-    if (!resourceModels.contains(item)) resourceModels[item] = new ProtoModel(typeMessage, nullptr);
-    auto resourceModel = resourceModels[item];
+    ProtoModel *res = resourceMap->GetResourceByName(item->type_case(), item->name());
+    BaseEditor *editor = factoryFunction->second(res, this);
 
-    QWidget *editor = factoryFunction->second(resourceModel, nullptr);
-    resourceModel->setParent(editor);
+    connect(editor, &BaseEditor::ResourceRenamed, resourceMap, &ResourceModelMap::ResourceRenamed);
+    connect(editor, &BaseEditor::ResourceRenamed, [=]() { treeModel->dataChanged(QModelIndex(), QModelIndex()); });

```

Comment:
I *like* this.

Reply:
Which part of it? If you mean the factory, that was me, he tacked on this resource renaming stuff. That said, I think there's a mistake here. Where he passes `QModelIndex()` to fire `dataChanged` I believe that's saying that the entire tree model changed when really only a single node did.

Reply:
I like the simple way in which you have registered a handler for a resource rename.